File: backend/app/infra/keychain.py
from __future__ import annotations

import secrets as secret_token
from dataclasses import dataclass
from typing import Protocol

import keyring


@dataclass(frozen=True)
class SecretLease:
    """表示仅在外部调用边界短时持有的明文凭据。"""
    value: str

    def __repr__(self) -> str:
        """返回不包含凭据明文的调试表示。"""
        return "SecretLease(value=[REDACTED])"


@dataclass(frozen=True)
class CredentialCheckResult:
    """表示凭据引用是否可用，但不返回凭据明文。"""
    available: bool
    secret_ref: str
    message: str


class CredentialAdapter(Protocol):
    """抽象 OS Keychain 与测试替身的凭据访问契约。"""
    async def save(self, provider: str, secret: str) -> str:
        """保存明文并返回安全引用。"""
        ...

    async def read(self, secret_ref: str) -> SecretLease:
        """按引用短时读取明文租约。"""
        ...

    async def delete(self, secret_ref: str) -> None:
        """删除指定凭据引用。"""
        ...

    async def check(self, secret_ref: str) -> CredentialCheckResult:
        """检查凭据可用性而不返回明文。"""
        ...


class MemoryCredentialAdapter:
    """仅供测试使用的内存凭据适配器，避免测试写入真实 Keychain。"""
    def __init__(self) -> None:
        """创建只在当前测试进程内保存数据的凭据映射。"""
        self._values: dict[str, str] = {}

    async def save(self, provider: str, secret: str) -> str:
        """保存测试凭据并返回不可推断明文的引用。"""
        secret_ref = f"memory://{provider}/{secret_token.token_urlsafe(16)}"
        self._values[secret_ref] = secret
        return secret_ref

    async def read(self, secret_ref: str) -> SecretLease:
        """按引用读取测试凭据。"""
        return SecretLease(self._values[secret_ref])

    async def delete(self, secret_ref: str) -> None:
        """删除测试凭据。"""
        self._values.pop(secret_ref, None)

    async def check(self, secret_ref: str) -> CredentialCheckResult:
        """检查引用是否存在，只返回状态和脱敏消息。"""
        available = secret_ref in self._values
        return CredentialCheckResult(
            available=available,
            secret_ref=secret_ref,
            message="Credential is available" if available else "Credential is not bound",
        )

    def clear(self) -> None:
        """清空测试凭据，避免测试之间共享秘密。"""
        self._values.clear()


class KeyringCredentialAdapter:
    """使用操作系统 Keychain/Secret Service 保存生产凭据。"""

    # 凭据服务名固定，避免同一应用在系统凭据库中产生多个命名空间。
    SERVICE = "digital-harness-runtime"

    async def save(self, provider: str, secret: str) -> str:
        """将明文写入 OS Keychain，并仅返回 secretRef。"""
        secret_ref = f"keyring://{provider}/{secret_token.token_urlsafe(16)}"
        keyring.set_password(self.SERVICE, secret_ref, secret)
        return secret_ref

    async def read(self, secret_ref: str) -> SecretLease:
        """在调用边界按引用短时读取 Keychain 明文。"""
        value = keyring.get_password(self.SERVICE, secret_ref)
        if value is None:
            raise KeyError(secret_ref)
        return SecretLease(value)

    async def delete(self, secret_ref: str) -> None:
        """删除 Keychain 中的凭据；不存在时保持幂等。"""
        try:
            keyring.delete_password(self.SERVICE, secret_ref)
        except keyring.errors.PasswordDeleteError:
            return

    async def check(self, secret_ref: str) -> CredentialCheckResult:
        """检查 Keychain 引用是否可读，不暴露凭据内容。"""
        available = keyring.get_password(self.SERVICE, secret_ref) is not None
        return CredentialCheckResult(
            available=available,
            secret_ref=secret_ref,
            message="Credential is available" if available else "Credential is not bound",
        )

File: backend/src/infra/keychain.ts
import { randomUUID } from "node:crypto";

/** 外部调用边界的短时凭据租约；toString 不暴露明文。 */
export class SecretLease {
  constructor(readonly secret: string) {}
  /** 防止诊断或日志意外打印密钥正文。 */
  toString(): string { return "[SecretLease REDACTED]"; }
  /** 防止 console.inspect 意外打印密钥正文。 */
  [Symbol.for("nodejs.util.inspect.custom")](): string { return this.toString(); }
}
/** 不返回明文的凭据检查结果。 */
export type CredentialCheckResult = { available: boolean; message: string };
/** OS Keychain 的统一 TypeScript 适配边界。 */
export interface CredentialAdapter { save(provider: string, secret: string): Promise<string>; read(secretRef: string): Promise<SecretLease>; delete(secretRef: string): Promise<void>; check(secretRef: string): Promise<CredentialCheckResult>; }

/** 测试和本地诊断使用的内存凭据适配器；明文只存在进程内。 */
export class MemoryCredentialAdapter implements CredentialAdapter {
  private readonly values = new Map<string, { provider: string; secret: string }>();
  /** 保存凭据并只返回不可推断的引用。 */
  async save(provider: string, secret: string): Promise<string> { const ref = `memory://${provider}/${randomUUID()}`; this.values.set(ref, { provider, secret }); return ref; }
  /** 在实际调用边界读取短时租约。 */
  async read(secretRef: string): Promise<SecretLease> { const item = this.values.get(secretRef); if (!item) throw new Error("credential unavailable"); return new SecretLease(item.secret); }
  /** 删除进程内凭据。 */
  async delete(secretRef: string): Promise<void> { this.values.delete(secretRef); }
  /** 只返回是否存在且非空。 */
  async check(secretRef: string): Promise<CredentialCheckResult> { const item = this.values.get(secretRef); return item?.secret ? { available: true, message: "credential available" } : { available: false, message: "credential unavailable" }; }
  /** 清空测试替身中的全部明文。 */
  clear(): void { this.values.clear(); }
}

/** 使用系统命令访问 macOS Keychain；其他平台以明确阻断替代静默落盘。 */
export class SystemCredentialAdapter implements CredentialAdapter {
  private readonly service: string;
  /** 绑定仅供 sidecar 使用的 Keychain service 名称。 */
  constructor(service = "digital-harness") { this.service = service; }
  /** 保存到 OS Keychain 并返回 secretRef；密钥不进入 SQLite。 */
  async save(provider: string, secret: string): Promise<string> {
    const ref = `keychain://${provider}/${randomUUID()}`;
    if (process.platform !== "darwin") throw new Error("OS Keychain adapter requires a platform implementation");
    const { execFile } = await import("node:child_process");
    await new Promise<void>((resolve, reject) => execFile("security", ["add-generic-password", "-a", ref, "-s", this.service, "-w", secret, "-U"], (error) => error ? reject(new Error("OS Keychain save failed")) : resolve()));
    return ref;
  }
  /** 从 OS Keychain 读取短时 SecretLease。 */
  async read(secretRef: string): Promise<SecretLease> { const value = await this.runSecurity(["find-generic-password", "-a", secretRef, "-s", this.service, "-w"]); return new SecretLease(value.trim()); }
  /** 从 OS Keychain 删除指定引用。 */
  async delete(secretRef: string): Promise<void> { await this.runSecurity(["delete-generic-password", "-a", secretRef, "-s", this.service]); }
  /** 只探测引用是否可读取，不返回明文。 */
  async check(secretRef: string): Promise<CredentialCheckResult> { try { await this.read(secretRef); return { available: true, message: "credential available" }; } catch { return { available: false, message: "credential unavailable" }; } }
  private async runSecurity(args: string[]): Promise<string> { if (process.platform !== "darwin") throw new Error("OS Keychain adapter requires a platform implementation"); const { execFile } = await import("node:child_process"); return new Promise((resolve, reject) => execFile("security", args, { encoding: "utf8" }, (error, stdout) => error ? reject(new Error("OS Keychain operation failed")) : resolve(stdout))); }
}
